central/virtualmachine/datastore/telemetry.go

```diff
@@ -0,0 +1,104 @@
+	"github.com/stackrox/rox/pkg/sac/resources"
+	"github.com/stackrox/rox/pkg/set"
+	"github.com/stackrox/rox/pkg/telemetry/phonehome"
+)
+
+const (
+	// activeAgentScanRecencyThreshold defines how recent a scan must be to consider
+	// the VM agent as active. This threshold is included in the telemetry metric name.
+	activeAgentScanRecencyThreshold = 24 * time.Hour
+
+	// Metric name constants to ensure consistency between implementation and tests
+	metricClustersWithVMs     = "Total Secured Clusters With Virtual Machines"
+	metricTotalVMs            = "Total Virtual Machines"
+	metricVMsWithActiveAgents = "Total Virtual Machines With Active Agents (Last 24h)"
+)
+
+var (
+	log = logging.LoggerForModule()
+)
+
+// Gather returns a function that collects telemetry about virtual machines.
+// It tracks three metrics:
+// 1. Number of distinct secured clusters with at least one running VM
+// 2. Total number of virtual machines
+// 3. Number of VMs with active agents (received IndexReport within last 24 hours)
+//
+// When the ROX_VIRTUAL_MACHINES feature flag is disabled, this function returns
+// an empty map without performing any database queries, ensuring no performance impact.
+func Gather(ds DataStore) phonehome.GatherFunc {
+	return gatherWithTime(ds, time.Now)
+}
+
+// gatherWithTime allows injecting a time function for deterministic testing.
+func gatherWithTime(ds DataStore, nowFunc func() time.Time) phonehome.GatherFunc {
+	return func(ctx context.Context) (map[string]any, error) {
+		// Early return if virtual machines feature is disabled - zero performance impact
+		if !features.VirtualMachines.Enabled() {
+			return map[string]any{}, nil
+		}
+
+		// Use elevated permissions for telemetry gathering
+		ctx = sac.WithGlobalAccessScopeChecker(ctx,
+			sac.AllowFixedScopes(
+				sac.AccessModeScopeKeys(storage.Access_READ_ACCESS),
+				sac.ResourceScopeKeys(resources.VirtualMachine),
+			),
+		)
+
+		// Process VMs one-by-one without loading all into memory
+		clusterIDsWithRunningVMs := set.NewStringSet()
+		totalVMs := 0
+		vmsWithActiveAgents := 0
+		now := nowFunc()
+
+		err := ds.Walk(ctx, func(vm *storage.VirtualMachine) error {
+			totalVMs++
+
+			// Count VMs with active agents (scan received within threshold)
+			if scan := vm.GetScan(); scan != nil {
+				scanTime, err := protocompat.ConvertTimestampToTimeOrError(scan.GetScanTime())
+				if err != nil {
+					log.Debugf("Virtual machine %s has invalid scan_time: %v", vm.GetId(), err)
+				} else if now.Sub(scanTime) <= activeAgentScanRecencyThreshold {
+					vmsWithActiveAgents++
+				}
+			}
+
+			// Count clusters with RUNNING virtual machines
+			if vm.GetState() == storage.VirtualMachine_RUNNING {
+				clusterID := vm.GetClusterId()
+				if clusterID == "" {
+					// Log empty cluster IDs at debug level for troubleshooting
+					log.Debugf("Virtual machine %s has empty cluster_id", vm.GetId())
+					return nil
+				}
+				clusterIDsWithRunningVMs.Add(clusterID)
```

Comment:
nit: this is fine, considering that this place in the code is currently the end of the VM processing. Nevertheless, the early return could lead to problems later on if new checks on the VM (new metrics) are added after that code.
```suggestion
				if clusterID == "" {
					// Log empty cluster IDs at debug level for troubleshooting
					log.Debugf("Virtual machine %s has empty cluster_id", vm.GetId())
				} else {
					clusterIDsWithRunningVMs.Add(clusterID)
				}
```

Reply:
Done [f760870](https://github.com/stackrox/stackrox/pull/17167/commits/f760870e7584faa01e254ff344afbdbd9def7d05)